src/server/gc_worker/compaction_runner.rs

```diff
@@ -0,0 +1,728 @@
+                        |_| {}
+                    );
+                    candidates
+                }
+                Err(e) => {
+                    error!("failed to collect compaction candidates: {:?}", e);
+                    if self.sleep_or_stop(check_interval) {
+                        debug!("compaction-runner stopped");
+                        break;
+                    }
+                    continue;
+                }
+            };
+
+            if candidates.is_empty() {
+                info!("no compaction candidates found, sleeping");
+                if self.sleep_or_stop(check_interval) {
+                    debug!("compaction-runner stopped");
+                    break;
+                }
+                continue;
+            }
+
+            // Compact the candidates
+            let elapsed = match self.compact_candidates(candidates, &config) {
+                Some(elapsed) => elapsed,
+                None => {
+                    debug!("compaction-runner stopped");
+                    break;
+                }
+            };
+
+            // Sleep for remaining time in check interval, or start next round immediately
+            if elapsed < check_interval {
+                let remaining_sleep = check_interval - elapsed;
+                if self.sleep_or_stop(remaining_sleep) {
+                    debug!("compaction-runner stopped");
+                    break;
+                }
+            }
+        }
+    }
+
+    /// Collects all compaction candidates from all regions
+    fn collect_compaction_candidates(
+        &self,
+        gc_safe_point: u64,
+        config: &GcConfig,
+    ) -> Result<Vec<CompactionCandidate>> {
+        // Calculate heap capacity based on check duration (assuming 1 sec per
+        // compaction)
+        let check_duration_secs = config.auto_compaction.check_interval.as_secs() as usize;
+        let heap_capacity = check_duration_secs.max(10); // At least 10 candidates
+
+        debug!(
+            "collecting compaction candidates with heap capacity: {}",
+            heap_capacity
+        );
+
+        // Use a min-heap to keep top candidates (using Reverse for min-heap behavior)
+        // We use a heap to retain only the top candidates, which helps limit memory
+        // usage in case we need to track compaction scores for all regions.
+        // This module assumes that each compaction takes at least 1 second, ensuring
+        // full utilization of the time window between check intervals.
+        let mut candidates_heap: BinaryHeap<Reverse<CompactionCandidate>> =
+            BinaryHeap::with_capacity(heap_capacity);
+        let mut current_key = Key::from_encoded(b"".to_vec());
+
+        while let (Some(region), next_key) = self.get_next_region_context(current_key) {
+            // Evaluate this region as a compaction candidate
+            match self.evaluate_range_candidate(
```

Comment:
How long will it takes to evaluate a big instance, say 30k regions with 200k SST files in total? Will it takes too long?